diff --git a/block_visibility_groups.info.yml b/block_visibility_groups.info.yml
index 2cae6e3..af4fd9b 100644
--- a/block_visibility_groups.info.yml
+++ b/block_visibility_groups.info.yml
@@ -2,6 +2,7 @@ name: Block Visibility Groups
 type: module
 description: Allows the site administrator to easily manage complex visibility settings that apply to any block placed in a visibility group.
 core: 8.x
+core_version_requirement: ^8 || ^9
 package: Other
 dependencies:
   - drupal:block
diff --git a/block_visibility_groups_admin/src/Tests/GroupListerTest.php b/block_visibility_groups_admin/src/Tests/GroupListerTest.php
index 9e060cb..5d0e6a7 100644
--- a/block_visibility_groups_admin/src/Tests/GroupListerTest.php
+++ b/block_visibility_groups_admin/src/Tests/GroupListerTest.php
@@ -2,14 +2,14 @@
 
 namespace Drupal\block_visibility_groups_admin\Tests;
 
-use Drupal\simpletest\WebTestBase;
+use Drupal\Tests\BrowserTestBase;
 
 /**
  * Provides automated tests for the block_visibility_groups_admin module.
  *
  * @group block_visibility_groups_admin
  */
-class GroupListerTest extends WebTestBase {
+class GroupListerTest extends BrowserTestBase {
 
   /**
    * Drupal\block_visibility_groups\GroupEvaluator definition.
diff --git a/src/BlockVisibilityGroupedListBuilder.php b/src/BlockVisibilityGroupedListBuilder.php
index 1e999ce..50ba850 100644
--- a/src/BlockVisibilityGroupedListBuilder.php
+++ b/src/BlockVisibilityGroupedListBuilder.php
@@ -77,10 +77,10 @@ class BlockVisibilityGroupedListBuilder extends BlockListBuilder {
   public static function createInstance(ContainerInterface $container, EntityTypeInterface $entity_type) {
     return new static(
       $entity_type,
-      $container->get('entity.manager')->getStorage($entity_type->id()),
+      $container->get('entity_type.manager')->getStorage($entity_type->id()),
       $container->get('theme.manager'),
       $container->get('form_builder'),
-      $container->get('entity.manager')->getStorage('block_visibility_group'),
+      $container->get('entity_type.manager')->getStorage('block_visibility_group'),
       $container->get('state'),
       $container->get('messenger')
     );
diff --git a/src/Plugin/Condition/ConditionGroup.php b/src/Plugin/Condition/ConditionGroup.php
index d77b711..acd2f3a 100644
--- a/src/Plugin/Condition/ConditionGroup.php
+++ b/src/Plugin/Condition/ConditionGroup.php
@@ -146,7 +146,7 @@ class ConditionGroup extends ConditionPluginBase implements ContainerFactoryPlug
   public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
     $container->get('block_visibility_groups.group_evaluator');
     return new static(
-      $container->get('entity.manager')->getStorage('block_visibility_group'),
+      $container->get('entity_type.manager')->getStorage('block_visibility_group'),
       $container->get('plugin.manager.condition'),
       $container->get('request_stack')->getCurrentRequest(),
       $container->get('block_visibility_groups.group_evaluator'),
diff --git a/src/Tests/BlockVisibilityGroupsTestBase.php b/src/Tests/BlockVisibilityGroupsTestBase.php
index 85a05db..b67453a 100644
--- a/src/Tests/BlockVisibilityGroupsTestBase.php
+++ b/src/Tests/BlockVisibilityGroupsTestBase.php
@@ -7,12 +7,12 @@
 
 namespace Drupal\block_visibility_groups\Tests;
 
-use Drupal\simpletest\WebTestBase;
+use Drupal\Tests\BrowserTestBase;
 
 /**
  *
  */
-abstract class BlockVisibilityGroupsTestBase extends WebTestBase {
+abstract class BlockVisibilityGroupsTestBase extends BrowserTestBase {
 
   /**
    * Modules to enable.
